### ISSUE TITLE
Weighting loss for imbalanced classes

### ISSUE THREAD

**Issue by Nuccy90** [flairNLP/flair#413]
Hi,

I'm working with a dataset that has extremely imbalanced classes. I get a really low recall for the minority class, so I would need to weight the loss function to penalize not labelling the minority class correctly. Is there any straightforward way to do this, for example by using the PyTorch `weight` argument?

Thank you!

**Comment by alanakbik** [comment]
Hi @Nuccy90 that is interesting - we have not yet worked with reweighing losses but this seems like a feature that would be great to have in Flair. It would be interesting to know if there's a good PyTorch way to do this. 

I'm adding a feature tag to this issue - though I am not sure when we can get around to looking into this. So any help would be greatly appreciated!

**Comment by romanovzky** [comment]
I'm interested in this as well. As far as I know, the loss has to be able to take in class-weights and use them when before reducing mean the total loss. For example, see TF and Keras implementation at the `fit` step [here](https://github.com/tensorflow/tensorflow/issues/10021), [here](https://www.tensorflow.org/api_docs/python/tf/nn/weighted_cross_entropy_with_logits), and [here](https://keras.io/models/model/#fit).

The logic is the same to any ML optimiser that accepts loss contributions from distinct classes (example, [decision trees](https://scikit-learn.org/stable/modules/generated/sklearn.utils.class_weight.compute_class_weight.html), etc as can be seen in [sklearn-api](https://scikit-learn.org/stable/modules/generated/sklearn.utils.class_weight.compute_class_weight.html).

In addition, I'd suggest the inclusion of [Focal Loss](https://arxiv.org/pdf/1708.02002.pdf), which is an adaption of these ideas that has yielded benefits to imagenet benchmkars, for example.

**Comment by gccome** [comment]
This feature would be a great addition to Flair!

**Comment by pacocp** [comment]
In Pytorch you can use the weight argument in the loss function to assign weights to each class. An example of this can be found in the documentation for the CrossEntropyLoss: https://pytorch.org/docs/stable/nn.html#torch.nn.CrossEntropyLoss

Hope it helps!

**Comment by alanakbik** [comment]
This functionality was added in #1345 and is part of the current Flair release.
